src/connection.h

```diff
@@ -138,6 +153,8 @@ struct connection {
     ConnectionCallbackFunc conn_handler;
     ConnectionCallbackFunc write_handler;
     ConnectionCallbackFunc read_handler;
+    char *saddr;
+    char *daddr;
```

Comment:
have you considered embedding these two fields directly in `connection`?  the downside is extra 30 bytes for ipv4 (actually less because there would be no pointers needed and rounding might work in our favor) but then we save two allocations/deallocations and 2 pointers.